provider/datagen/src/driver.rs

```diff
@@ -0,0 +1,566 @@
+        Self {
+            keys: keys.into_iter().collect(),
+            ..self
+        }
+    }
+
+    /// Sets the fallback type that the data should be generated for. If locale fallback is
+    /// used at runtime, smaller data can be generated.
+    pub fn with_fallback_mode(self, fallback: FallbackMode) -> Self {
+        Self { fallback, ..self }
+    }
+
+    /// Sets the locales to generate.
+    pub fn with_locales(self, locales: impl IntoIterator<Item = LanguageIdentifier>) -> Self {
+        Self {
+            locales: Some(locales.into_iter().collect()),
+            ..self
+        }
+    }
+
+    /// Sets this driver to generate all available locales.
+    pub fn with_all_locales(self) -> Self {
+        Self {
+            locales: None,
+            ..self
+        }
+    }
+
+    /// By default, the collations `big5han`, `gb2312`, and those starting with `search`
+    /// are excluded. This method can be used to reennable them.
+    ///
+    /// The special string `"search*"` causes all search collation tables to be included.
+    pub fn with_collations(self, collations: impl IntoIterator<Item = String>) -> Self {
+        Self {
+            collations: collations.into_iter().collect(),
+            ..self
+        }
+    }
+
+    /// Sets this driver to generate the given segmentation models, to the extent required by the
+    /// chosen data keys.
+    ///
+    /// The currently supported dictionary models are
+    /// * `cjdict`
+    /// * `burmesedict`
+    /// * `khmerdict`
+    /// * `laodict`
+    /// * `thaidict`
+    ///
+    /// The currently supported LSTM models are
+    /// * `Burmese_codepoints_exclusive_model4_heavy`
+    /// * `Khmer_codepoints_exclusive_model4_heavy`
+    /// * `Lao_codepoints_exclusive_model4_heavy`
+    /// * `Thai_codepoints_exclusive_model4_heavy`
+    ///
+    /// If a model is not included, the resulting line or word segmenter will apply rule-based
+    /// segmentation when encountering text in a script that requires the model, which will be
+    /// incorrect.
+    ///
+    /// If multiple models for the same language and segmentation type (dictionary/LSTM) are
+    /// listed, the first one will be used.
+    pub fn with_segmenter_models(self, models: impl IntoIterator<Item = String>) -> Self {
+        Self {
+            segmenter_models: models.into_iter().collect(),
+            ..self
+        }
+    }
+
+    /// Exports data from the given provider to the given exporter.
+    ///
+    /// See
+    /// [`BlobExporter`](icu_provider_blob::export),
+    /// [`FileSystemExporter`](icu_provider_fs::export),
+    /// and [`BakedExporter`](crate::baked_exporter).
+    pub fn export(
+        self,
+        provider: &(impl IterableDynamicDataProvider<ExportMarker>
+              + DataProvider<LocaleFallbackLikelySubtagsV1Marker>
+              + DataProvider<LocaleFallbackParentsV1Marker>
+              + DataProvider<CollationFallbackSupplementV1Marker>
```

Comment:
Yes, but also no, this doesn't really fit our unstable/any/buffer constructor pattern, because the provider itself is dynamic. 

We could define a trait that can construct a `LocaleFallbacker` and is a `IterableDynamicDataProvider<ExportMarker>`.

Reply:
I'm fine with the trait bound; we just need to make clear that the trait bound is subject to change.

Reply:
It shouldn't be needed, we should just be able to access it through `ExportableProvider`. I've played around with getting `Any`s but couldn't come up with a quick solution.

Reply:
`trait ExportableDataPayload` has a function `as_any` so it should be possible to downcast a `DataPayload<ExportMarker>` without any code changes. Something like

```rust
payload.get().payload.as_any().downcast_ref::<LikelySubtagsV1<'static>>()
```

(okay, maybe `ExportBox` needs to get a function since its field `payload` is private)